File: C++/virtual-world-simulator/src/Entity/Animals/Fox.cpp
#include "Fox.h"

Fox::Fox(World* world, Point position) :
    Animal{ AssetManager::getAssetManager()->getAsset("fox"), world, position, 'F', 3, 7 }
{
}

Fox::Fox(World* world, Point position, int lifeSpan, int strength) :
    Animal{ AssetManager::getAssetManager()->getAsset("fox"), world, position, 'F', strength, 7, lifeSpan }
{
}

void Fox::update(std::fstream& logFile)
{
    int currentX = this->position.x;
    int currentY = this->position.y;

    std::vector<Point> possibleMoves;
    Entity*** map = world->getMap();

    for (int i = currentX - 1; i <= currentX + 1; i++)
    {
        if (i < 0 || i > world->getHeight() - 1)
            continue;

        for (int j = currentY - 1; j <= currentY + 1; j++)
        {
            if (j < 0 || j > world->getWidth() - 1 ||
                (i == currentX && j == currentY))
                continue;

            if (map[i][j] == nullptr || map[i][j]->getStrength() <= this->strength)
                possibleMoves.push_back(Point(i, j));
		}
    }
    if(possibleMoves.size() > 0)
    {
        Point newPosition = possibleMoves[Utility::random(0, possibleMoves.size() - 1)];
        this->move(logFile, newPosition);
    }
    Entity::update(logFile);
}

bool Fox::collision(std::fstream& logFile, Entity& entity)
{
    Fox* fox = dynamic_cast<Fox*>(&entity);
    if (fox != nullptr)
    {
        this->breed<Fox>(logFile, *fox);
        return false;
    }

    return Animal::collision(logFile, entity);
}

File: Python/virtual-world-simulator/src/Entity/Animals/Fox.py
import random

import pygame
from src.Entity.Animals.Animal import Animal


class Fox(Animal):

    def __init__(self, world, position):
        temp = pygame.image.load('assets/fox.png')
        scaled = pygame.transform.scale(temp, (world.scale, world.scale))
        super().__init__(scaled, world, position, 'F', 3, 7)

    def collision(self, entity):
        if isinstance(entity, Fox):
            # print("BREED")
            return False

        return super().collision(entity)

    def update(self):
        currentX = self._position[0]
        currentY = self._position[1]

        possibleMoves = []

        for i in range(currentX - 1, currentX + 2):
            if i < 0 or i > self._world.height - 1:
                continue

            for j in range(currentY - 1, currentY + 2):
                if j < 0 or j > self._world.width - 1 or i == currentX or j == currentY:
                    continue

                element = self._world.getMapElement(i, j)
                if element is None or element.strength <= self._strength:
                    possibleMoves.append((i, j))

        numberOfPossibleMoves = len(possibleMoves)
        if numberOfPossibleMoves > 0:
            newPosition = possibleMoves[random.randint(0, numberOfPossibleMoves - 1)]
            self.move(newPosition)

        self._lifespan += 1
